docs/decisions/0006-pluggable-lti-configuration.rst

```diff
@@ -0,0 +1,128 @@
+
+#. Allow externally configured tools (a new {{CONFIG_EXTERNAL}} choice for the configuration storage).
+#. Two new fields to store an unique external tool id and the plugin name/location.
+#. Changes to the {{get_lti_consumer}} methods to allow instancing a {{LtiConsumer}} from externally managed configurations.
+
+
+The filter part will allow the XBlock LTI Consumer to query re-usable and/or plugabble LTI configurations and display them to course 
+creators as configuration options in Studio.
+
+The filter response should contain all the necessary information for the XBlock to locate and use the externally managed LTI configuration.
+The reponse **should** be structured this way:
+
+.. code-block:: json
+    {
+        "filter_name": {
+            "api_path": "plugin_1_module_name.api.get_lti_config",
+            "available_configurations": [
+                {
+                    "lti_version": "LTI_1P1",
+                    "display_name": "Reusable LTI tool 1",
+                    "lti_config_id": 2,
+                }
+                {
+                    "lti_version": "LTI_1P3",
+                    "display_name": "Reusable LTI tool 2 - LTI 1.3",
+                    "lti_config_id": 3,     
+                }
+            ]
+        }
+    }
+
+The LTI Configuration providers are responsible for restricting which configurations are available to each course/context. 
+
+Each LTI configuration provider MUST:
+
+* Provide a `get_lti_config` method through the `api_path` parameter conforming with the API parameters described in the `Plugin API interface`.
+* Store and manage LTI configurations, including removing `LtiConfiguration` entries when they are deleted from the plugin.
+
+lti-consumer Python APIs
+========================
+
+**`get_reusable_lti_configurations`**
+Used by the LTI Consumer XBlock to retrieve available re-usable LTI configurations that can be used in a block.
+This is gated by a feature flag so that the previous behavior is retained:
+
+
+.. list-table::
+   :widths: auto
+   :header-rows: 1
+
+   * - Toggle name
+     - Type
+     - Behavior
+   * - LTI_ENABLE_REUSABLE_CONFIGURATIONS
+     - Feature flag
+     - This flag enables LTI Configuration reusability and will display an extra field during the LTI 
+       configuration that will allow selection from pre-set configurations.
+
+
+
+Plugin API interface
+====================
+
+**`get_lti_config`**
+
+Each plugin must supply a method that returns the necessary LTI parameters settings so that the LTI Configuration 
+can instance one of the following LTI consumers from it:
+
+* :code:`lti_consumer.lti_1p1.consumer.LtiConsumer1p1` for LTI 1.1
+* :code:`lti_consumer.lti_1p3.consumer.LtiAdvantageConsumer` for LTI 1.3.
+
+A method in LTI Configuration will be responsible for instancing the LTI Consumer class and returning it to the XBlock.
+It'll also handle creating any acessory models to enable LTI Advange services. 
```

Comment:
I think for LTI1.1 at least we can just directly return this from the filter, we don't need to call `get_lti_config` since the filter will already know if there are any credentials configured for this block.

Instead of returning the configured consumer (which will involve each filter having a lot of the same boilerplate) we can just have the filter return a configuration object, which can then be passed to a function that builds the consumer from that.

Reply:
@xitij2000 I agree with the fact that there will be some boilerplate code in the filters, my initial goal was to reduce the complexity inside the core by removing the logic of configuring the LTI consumer from it (look at the LTI 1.3 method: https://github.com/edx/xblock-lti-consumer/blob/master/lti_consumer/models.py#L247, it's big).

Also, if we retrieve configuration/credentials directly using the filter instead of calling a filter method, we'll have to run the filter pipeline every time the LTI content is launched or one of it's APIs is used (which is less performant than calling a direct function and retrieving the configuration).

What to you think of: 
1. Removing the need for the plugins to return the LTI consumer (just return the credentials + any additional variables required) and leave the LTI consumer instancing to model method/helper.
2. Keep retrieving the credentials using a function name + ID instead of running the filter pipeline on each LTI launch.